hail/src/main/scala/is/hail/expr/types/physical/PBaseStruct.scala

```diff
@@ -248,5 +272,76 @@ abstract class PBaseStruct extends PType {
     }
   }
 
+  override def copyFromType(mb: MethodBuilder, region: Code[Region], srcPType: PType, srcStructAddress: Code[Long],
+    allowDowncast: Boolean = false, forceDeep: Boolean = false): Code[Long] = {
+    assert(srcPType.isInstanceOf[PBaseStruct])
+
+    val sourceType = srcPType.asInstanceOf[PBaseStruct]
+
+    assert(sourceType.size == this.size)
+
+    val sourceFundamentalTypes = sourceType.fields.map(_.typ.fundamentalType)
+    if(this.fields.map(_.typ.fundamentalType) == sourceFundamentalTypes) {
+      if(!forceDeep) {
+        return srcStructAddress
+      }
+
+      if(sourceFundamentalTypes.forall(_.isPrimitive)) {
```

Comment:
this is a bit wasteful, could implement in one loop, but I think this fits our typical style better